modules/imgproc/src/colormap.cpp

```diff
@@ -797,9 +797,29 @@ namespace colormap
             CV_Error(Error::StsAssert, "cv::LUT only supports tables of size 256.");
         if (userColor.type() != CV_8UC1 && userColor.type() != CV_8UC3)
             CV_Error(Error::StsAssert, "cv::LUT only supports tables CV_8UC1 or CV_8UC3.");
-        colormap::UserColorMap cm(userColor.getMat());
-
-        cm(src, dst);
+        if (src.isMat() && userColor.isMat() && ((src.type() == CV_8UC1) || (src.type() == CV_8UC3)))
+        {
+            const cv::Mat srcMat = src.getMat();
+            cv::Mat srcGray;
+            if (srcMat.channels() == 1)
+              srcGray = srcMat;
+            else
+              cv::cvtColor(srcMat, srcGray, cv::COLOR_BGR2GRAY);
+            cv::Mat userColorMat = userColor.getMat();
+            dst.create(src.size(), userColor.type());
+            cv::Mat dstMat = dst.getMat();
+            const size_t elemSize = userColorMat.elemSize();
+            srcGray.forEach<unsigned char>([&](unsigned char& pixel, const int* position) -> void {
+                const int row = position[0];
+                const int col = position[1];
+                memcpy(dstMat.data+row*dstMat.step+col*elemSize, userColorMat.data+pixel*userColorMat.step, elemSize);
```

Comment:
> `memcpy(..., elemSize)`

Provide performance numbers without and with your patch.

Reply:
Intel(R) Core(TM) i7-4790 CPU @ 3.60GHz
Windows 7 64b

```
//testing cv::applyColorMap
{
	cv::setRNGSeed(1234);
	cv::Mat colorMap = cv::Mat(256, 1, CV_8UC3);
	cv::randu(colorMap, 0, 255);
	std::vector<cv::Size> sizes = {{32, 32}, {64, 64}, {128, 128},  {256, 256}, {512, 512}, {1024, 1024}};
	for(const auto& size : sizes)
	{
		cv::Mat src(size, CV_8UC1);//CV_8UC1//CV_8UC3
		cv::Mat dst(size, CV_8UC3);

		__int64 totalTicks = 0;
		const size_t nbIters = 20000;
		for(size_t i = 0 ; i <nbIters ; ++i)
		{
			cv::randu(src, 0, 255);
			const auto ref = cv::getTickCount();
			cv::applyColorMap(src, dst, colorMap);
			const auto now = cv::getTickCount();
			totalTicks += (now-ref);
		}
		printf("%llu x cv::applyColorMap() for size(%d,%d) and type %s: %f ms\r\n",
			(unsigned long long)nbIters,
			size.width, size.height,
			(src.type() == CV_8UC1) ? "CV_8UC1" :
			(src.type() == CV_8UC3) ? "CV_8UC3" :
			"?",
			1000.*totalTicks/cv::getTickFrequency());
	}
	cv::waitKey();
}
```
results :

CV_8UC1
original code :
20000 x cv::applyColorMap() for size(32,32) and type CV_8UC1: 88.819806 ms
20000 x cv::applyColorMap() for size(64,64) and type CV_8UC1: 269.382595 ms
20000 x cv::applyColorMap() for size(128,128) and type CV_8UC1: 1035.405428 ms
20000 x cv::applyColorMap() for size(256,256) and type CV_8UC1: 3899.305621 ms
20000 x cv::applyColorMap() for size(512,512) and type CV_8UC1: 11115.692492 ms
20000 x cv::applyColorMap() for size(1024,1024) and type CV_8UC1: 39989.798904 ms

with patch :
20000 x cv::applyColorMap() for size(32,32) and type CV_8UC1: 105.990024 ms
20000 x cv::applyColorMap() for size(64,64) and type CV_8UC1: 178.348563 ms
20000 x cv::applyColorMap() for size(128,128) and type CV_8UC1: 390.924573 ms
20000 x cv::applyColorMap() for size(256,256) and type CV_8UC1: 1185.652095 ms
20000 x cv::applyColorMap() for size(512,512) and type CV_8UC1: 3715.295696 ms
20000 x cv::applyColorMap() for size(1024,1024) and type CV_8UC1: 15669.887541 ms

CV_8UC3
original code :
20000 x cv::applyColorMap() for size(32,32) and type CV_8UC3: 159.482952 ms
20000 x cv::applyColorMap() for size(64,64) and type CV_8UC3: 351.207949 ms
20000 x cv::applyColorMap() for size(128,128) and type CV_8UC3: 1232.039576 ms
20000 x cv::applyColorMap() for size(256,256) and type CV_8UC3: 7554.557066 ms
20000 x cv::applyColorMap() for size(512,512) and type CV_8UC3: 17448.084637 ms
20000 x cv::applyColorMap() for size(1024,1024) and type CV_8UC3: 69502.533269 ms

with patch :
20000 x cv::applyColorMap() for size(32,32) and type CV_8UC3: 145.020181 ms
20000 x cv::applyColorMap() for size(64,64) and type CV_8UC3: 263.634428 ms
20000 x cv::applyColorMap() for size(128,128) and type CV_8UC3: 690.046396 ms
20000 x cv::applyColorMap() for size(256,256) and type CV_8UC3: 1889.553916 ms
20000 x cv::applyColorMap() for size(512,512) and type CV_8UC3: 5617.506280 ms
20000 x cv::applyColorMap() for size(1024,1024) and type CV_8UC3: 22068.121624 ms